compiler/rustc_hir_analysis/src/lib.rs

```diff
@@ -127,20 +127,23 @@ fn require_c_abi_if_c_variadic(
     }
 
     let extended_abi_support = tcx.features().extended_varargs_abi_support();
-    let conventions = match (extended_abi_support, abi.supports_varargs()) {
+    let extern_system_varargs = tcx.features().extern_system_varargs();
+    let conventions = if extended_abi_support { CONVENTIONS_UNSTABLE } else { CONVENTIONS_STABLE };
+    match abi {
         // User enabled additional ABI support for varargs and function ABI matches those ones.
-        (true, true) => return,
+        ExternAbi::System { .. } if extern_system_varargs => return,
```

Comment:
Could you merge these guarded arms and the unguarded arms? I think:
```
ExternAbi::System { .. } if extern_system_varargs => return,
ExternAbi::System { .. } => {
    feature_err(&tcx.sess, sym::extern_system_varargs, span, UNSTABLE_EXPLAIN).emit();
}
```

Is much harder to read than:

```
ExternAbi::System { .. } => {
    if extern_system_varargs {
        return;
    } else {
        feature_err(&tcx.sess, sym::extern_system_varargs, span, UNSTABLE_EXPLAIN).emit();
}
```

Also, looking at the cotrol flow here... I think it's a bit strange here that we're issuing both a feature error and another error below. Maybe that could be consolidated/deduplicated too. Thoughts?

Reply:
hm, is c9c1ab40fa5aaec953124bdcba3e299c08811686 more readable to you?

it's pretty strange control flow, yes. I had been thinking about altering the entire way the error is computed, but it will work better after https://github.com/rust-lang/rust/pull/136901 lands (and I'm fine with delaying this until it does)

Reply:
my initial thought in the way I wrote it was to put all "escape paths" next to each other and then have all "error paths" next to each other.

Reply:
I tried to rewrite some things to hopefully make the thought process clearer